crates/ruff/src/rules/pylint/rules/redefined_loop_name.rs

```diff
@@ -0,0 +1,325 @@
+        match &stmt.node {
+            // For and async for.
+            StmtKind::For { target, .. } | StmtKind::AsyncFor { target, .. } => {
+                self.assignment_targets.extend(
+                    assignment_targets_from_expr(target, self.dummy_variable_rgx).map(|expr| {
+                        ExprWithBindingKind {
+                            expr,
+                            binding_kind: BindingKind::For,
+                        }
+                    }),
+                );
+            }
+            // With.
+            StmtKind::With { items, .. } => {
+                self.assignment_targets.extend(
+                    assignment_targets_from_with_items(items, self.dummy_variable_rgx).map(
+                        |expr| ExprWithBindingKind {
+                            expr,
+                            binding_kind: BindingKind::With,
+                        },
+                    ),
+                );
+            }
+            // Assignment, augmented assignment, and annotated assignment.
+            StmtKind::Assign { targets, .. } => {
+                self.assignment_targets.extend(
+                    assignment_targets_from_assign_targets(targets, self.dummy_variable_rgx).map(
+                        |expr| ExprWithBindingKind {
+                            expr,
+                            binding_kind: BindingKind::Assignment,
+                        },
+                    ),
+                );
+            }
+            StmtKind::AugAssign { target, .. } | StmtKind::AnnAssign { target, .. } => {
+                self.assignment_targets.extend(
+                    assignment_targets_from_expr(target, self.dummy_variable_rgx).map(|expr| {
+                        ExprWithBindingKind {
+                            expr,
+                            binding_kind: BindingKind::Assignment,
+                        }
+                    }),
+                );
+            }
+            _ => {}
+        }
+        // Decide whether to recurse.
+        match &stmt.node {
+            // Don't recurse into blocks that create a new scope.
+            StmtKind::ClassDef { .. } => {}
+            StmtKind::FunctionDef { .. } => {}
+            // Otherwise, do recurse.
+            _ => {
+                visitor::walk_stmt(self, stmt);
+            }
+        }
+    }
+}
+
+fn assignment_targets_from_expr<'a, U>(
+    expr: &'a Expr<U>,
+    dummy_variable_rgx: &'a HashableRegex,
+) -> Box<dyn Iterator<Item = &'a Expr<U>> + 'a> {
+    // The Box is necessary to ensure the match arms have the same return type - we can't use
+    // a cast to "impl Iterator", since at the time of writing that is only allowed for
+    // return types and argument types.
+    match &expr.node {
+        ExprKind::Attribute {
+            ctx: ExprContext::Store,
+            ..
+        } => Box::new(iter::once(expr)),
+        ExprKind::Subscript {
+            ctx: ExprContext::Store,
+            ..
```

Comment:
If I understand correctly, subscripts and attributes can never appear in the `with`, `for`, etc. bindings. So can these ever trigger a violation? Could we just omit them, and only include `ExprKind::Name` nodes here? That would simplify the `// Compute the textual name of the outer target lazily.` step below, since we could just use the name's `id` attribute rather than extracting from the source code.

Reply:
Surprisingly, they very much can, both for attributes and subscripts. I didn't know this either until I tried it! I have never seen this in real code, but it's valid Python, and it works:

```python
In [1]: class A:
   ...:     pass
   ...: a = A()

In [2]: for a.i in range(5):
   ...:     for a.i in range(1):  # PLW2901
   ...:         print(a.i)
   ...:
0
0
0
0
0

In [3]: a = {}

In [4]: for a["i"] in range(5):
   ...:     for a["i"] in range(1):  # PLW2901
   ...:         print(a["i"])
   ...:
0
0
0
0
0
```

Reply:
What?? You're kidding me.

Reply:
yes, see also https://github.com/charliermarsh/ruff/issues/2208!